### PR TITLE
Generate static website docs from swift components

### PR DESCRIPTION
**Pull request by gruberb**
This PR is a first step in building static-website docs for the Swift components produced by `application-services`. 

It adds a new folder inside `automation`. Inside, we build a new Swift project, with all the `.swift` and `.h` files from the output of `uniffi-bindgen`. We have to manipulate a few things to make it work, and provide a fixed `module.modulemap` and `Package.swift`.

In the end, we use `jazzy` to generate the documentation.

## Feedback

I decided to move this to this repo instead of `rust-components-swift`, because it feels closer to the source, and we really just need the output of `uniffi-bindgen` to build this. No `Xcframework` or `Xcodeproject` needed.

## Outcome (local)
Screenshot 2024-08-29 at 09 26 25

### CODE REVIEW

**Comment by mhammond** [comment]
We briefly discussed UDL yesterday, so it's worth pointing out that the fxa-client crate has [extensive docs in UDL](https://github.com/mozilla/application-services/blob/main/components/fxa-client/src/fxa_client.udl) which we should make sure are rendered. I believe some edits will be needed though as their is a mix of `///` and `//` comments there and IIRC, only triple slash docs are extracted.

File: automation/swift-components-docs/Cargo.toml
Comment (bendk) [inline automation/swift-components-docs/Cargo.toml:28]: Could you use a dependency on `megazord_ios` instead of this list?  That way the crate list will always be kept in sync.

File: automation/swift-components-docs/Sources/MozillaRustComponents/include/module.modulemap
Comment (bendk) [inline automation/swift-components-docs/Sources/MozillaRustComponents/include/module.modulemap:16]: We really should create a way for UniFFI to help here.   Maybe if you use library mode and generate swift files, there could be an option for creating a single modulemap with everything in it.  In the meantime, I'm totally fine with hacking it with a hard-coded list of headers.

File: automation/swift-components-docs/replace-imports.sh
Comment (bendk) [inline automation/swift-components-docs/replace-imports.sh:8]: I'm don't totally understand this one, but could this be avoided if UniFFI had support for generating a modulemap for all the components?

File: .github/workflows/swift-docs.yaml
Comment (bendk) [inline .github/workflows/swift-docs.yaml:22]: +1 to Jazzy.  If Glean is using it, then I love the choice.

**Review by bendk — COMMENTED** [review]
Nice!  I guess the main thing is figuring out the CI.  Is it possible to commit the HTML to the `gh-pages` branch?  I think that's where we eventually want this to live.

If possible, it would be nice to make this a step in the GH actions in that branch.  I guess we need to enable the Ruby setup action though?

For now, it's not necessary to integrate it with the rest of the docs.  It would just be great to have it somewhere on the static site.  Later on we can re-organize those docs and start linking to the API reference.

**Comment by gruberb** [comment]
> We briefly discussed UDL yesterday, so it's worth pointing out that the fxa-client crate has [extensive docs in UDL](https://github.com/mozilla/application-services/blob/main/components/fxa-client/src/fxa_client.udl) which we should make sure are rendered. I believe some edits will be needed though as their is a mix of /// and // comments there and IIRC, only triple slash docs are extracted.

I didn't want to include an upgrade to `uniffi = "0.28.1`, because I think that's a different issue. I can though if we want. I quickly tested it locally, and yes, the doc comments (`///`) are extracted and shown in the output docs:

Screenshot 2024-08-29 at 12 29 39

We can open a PR to upgrade `uniffi` and adjust all the UDL comments to be `///` instead of `//`? Can also do it in this PR.

File: automation/swift-components-docs/replace-imports.sh
Comment (gruberb) [reply automation/swift-components-docs/replace-imports.sh:8]: Yeah that's a weird one. At this point, I surrender to the Swift gods. I tried what feels like a hundred different setups until `spm` finally included the header files.

The issue, I believe, is, that the project is named `MozillaRustComponents`, and the `.swift` files try to import `MozillaRustComponents` as well, because that's the name of the generated `module.modulemap` file. It works for `Xcframework`, but it didn't work for this setup (pure Swift). 

I can play around with the namings and see if we can leave the imports as is...

File: automation/swift-components-docs/Sources/MozillaRustComponents/include/module.modulemap
Comment (gruberb) [reply automation/swift-components-docs/Sources/MozillaRustComponents/include/module.modulemap:16]: I was "fine" with hardcoding it, because we currently have to hardcode any new component in the `generate-swift-project.sh` as well. But as soon as your PR lands with the library mode, we should auto-generate this `modulemap` file as well. I can put together a small script which does it as well as of now, so we don't hardcode anything here.

**Comment by gruberb** [comment]
I requested to use `ruby/setup-ruby` for this repository:
https://bugzilla.mozilla.org/show_bug.cgi?id=1915656

**Comment by gruberb** [comment]
@bendk I simplified the `Cargo.toml` a lot, and we directly use the `megazords/ios-rust` dependencies directly.

We have one last challenge: Where do we host this? Different options.

* We can use subfolders in the `docs` branch. Which means we have to integrate this workflow with the `generate-docs` workflow. We then create a subfolder there called `swift-docs`, which we then can link.
* We create a new repository, called `uniffi-docs`, where we also host the topic guides, and push it there. 

The problem with both is: We cannot host multiple GitHub pages domains for a single repository.

File: megazords/ios-rust/Cargo.toml
Comment (gruberb) [inline megazords/ios-rust/Cargo.toml:9]: Had to add `rlib` so I can reference it inside `swift-components-docs`. This will increase the compile times.

**Comment by bendk** [comment]
>   We can use subfolders in the `docs` branch. Which means we have to integrate this workflow with the `generate-docs` workflow. We then create a subfolder there called `swift-docs`, which we then can link.

I like this option.  We only want one static site in the end, so we may as well put everything in one branch.  You could either adjust this PR or do that in a follow up, either way is fine with me.  I think the branch is called `gh-pages` though right?

**Review by bendk — APPROVED** [review]
Looks great to me, as you mention the only remaining issue is where to put the docs.  I'll let you decide if you want to adjust this PR or create a follow-up.

**Pull request merged**
